src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+
+            // don't repeat this for subsequent EVs
+            if (evindex < 2) 
+            {
+              module_config->writeEvent(index, &msg->data[1]);
+            }
+
+            module_config->writeEventEV(index, evindex, evval);
+
+            // recreate event hash table entry
+            // DEBUG_SERIAL << F("> updating hash table entry for idx = ") << index << endl;
+            module_config->updateEvHashEntry(index);
+
+            // respond with WRACK
+            controller->sendWRACK();  // Deprecated in favour of GRSP_OK
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), GRSP_OK);
+          } 
+          else 
+          {
+            // DEBUG_SERIAL << F("> no free event storage, index = ") << index << endl;
+            // respond with CMDERR & GRSP
+            controller->sendCMDERR(CMDERR_TOO_MANY_EVENTS);
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_TOO_MANY_EVENTS);
+          }
+        }
       }
 
-      // if existing or new event space found, write the event data
-      if (index < module_config->EE_MAX_EVENTS)
-      {
-        // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
-        // DEBUG_SERIAL << F("> writing EV = ") << evindex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+      return PROCESSED;
 
-        // don't repeat this for subsequent EVs
-        if (evindex < 2)
+    case OPC_EVLRNI:
+      // F5 - learn an event using event index
+
+      DEBUG_SERIAL << endl
+                   << F("> EVLRNI for source nn = ") << nn << endl;
+
+      // we must be in learn mode
+      if (bLearn) 
+      {
+        if (msg->len < 8) 
+        {
+          controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_CMD);
+          controller->sendCMDERR(CMDERR_INV_CMD);
+        } 
+        else 
         {
-          module_config->writeEvent(index, &msg->data[1]);
+          byte index = msg->data[5];
+          byte evIndex = msg->data[6];
+          byte evVal = msg->data[7];
+
+          // invalid index
+          if (index >= module_config->EE_MAX_EVENTS) 
+          {
+            DEBUG_SERIAL << F("> invalid index") << endl;
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_EV_IDX);
+            return PROCESSED;
+          }
+          // write the event data
+          else 
+          {
+            // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
+            DEBUG_SERIAL << F("> writing EV = ") << evIndex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+
+            // don't repeat this for subsequent EVs
+            if (evIndex < 2) 
+            {
+              module_config->writeEvent(index, &msg->data[1]);
```

Comment:
This needs more explaining. What are we writing here?  Why are we writing this only for the first event variable?

Reply:
The event will be there because otherwise it won't have an index.  I am using this opcode to teach producers short events.  It can also be used to allow a producer to spoof.  I suppose the other way would be to check if the NN and the EN in the event table match the incoming for all instances of the op code and change the table if not.  I'll look at this as possible enhancement.

Reply:
Now revised how NN & EN are checked and updated as necessary.  This code compiles but I haven't been able to fully test it yet.  Is there any reason why this shouldn't run OK?

Reply:
No, "if (eventTableNNEN != &msg->data[1])" won't work. This is comparing addresses to the data. To compare the four bytes you should write this as "if (memcmp(eventTableNNEN, &msg->data[1], 4) == 0)"
"memcpy" works similar to "strcmp" that I covered in the talk about strings, but it works off memory so doesn't check for terminating null byte. Instead you need the "4" as a length of memory to be compared.

Reply:
Ah, thanks Sven.  Thought it looked too simple!  Now corrected